### DISCUSSION TITLE
multiple schema

### DISCUSSION THREAD

**Discussion by inikalaev** [skryukov/skooma#14] [Q&A] [answered]
Hey!
 How can I test multiple circuits at the same time? I have an admin schema and a public schema.

I tried using a construction of the form:
```ruby
  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema, path_prefix: "/admin/v1"], type: :request

  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema, path_prefix: "/public/v1"], type: :request
```
But then the prefix does not work correctly

I tried to create separate helpers, but the prefix and endpoint of the form /test/test/{id} not found in schema

```ruby
# admin_helper.rb
  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema, path_prefix: "/admin/v1"], type: :request
```

```ruby 
# public_helper.rb
  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema, path_prefix: "/public/v1"], type: :request
```

**Comment by skryukov** [comment, marked as answer]
_In the discussion to the initial answer we found out an issue in `skooma`, so here is an updated answer:_

I'd suggest using RSpec flags:

```ruby
  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema, path_prefix: "/admin/v1"], :admin_api
  config.define_derived_metadata(file_path: %r{/spec/requests/admin}) { |metadata| metadata[:admin_api] = true }

  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema, path_prefix: "/public/v1"], :public_api
  config.define_derived_metadata(file_path: %r{/spec/requests/public}) { |metadata| metadata[:public_api] = true }
```

Full example is now available here: https://github.com/skryukov/skooma/tree/main/examples/rails_app

**Reply by inikalaev** [reply]
Not work for me :(. Path not found.

**Reply by skryukov** [reply]
🤔 I wrote `file_path: %r{/spec/requests/public}` and `file_path: %r{/spec/requests/admin}` matchers, but that's just a guess, is your specs dir structure looks like that?

If you store both admin and public specs in the same dir and there is no way to automatically infer metadata, you can drop `config.define_derived_metadata` lines and just set metadata manually in every spec (but that's kinda painful): 
```ruby
describe "/materials", :public_api, do
  #...
end
```

**Reply by inikalaev** [reply]
No They are in different folders, but it does not work(. I tried to set it for each test, but it doesn't work either. Could there be a problem with the prefix?

**Reply by skryukov** [reply]
Hmm, this should not be related to the `skooma`, it's more on the RSpec side. Make sure that you put both `config.include ...` in the `rails_helper.rb` and then include `:public_api` or `:admin_api` to a related `describe "..."`. Also make sure that `type` is set to `:request`.

If you're uncertain if `skooma` does something funny, you can make smth like this (didn't test this snippet, so take this with a grain of salt):

```ruby
# spec/rails_helper.rb

MyHelper = Module.new do
  def foo
    puts "foo"
  end
end
# ...

RSpec.configure do |config|
  # ...
  config.include MyHelper, :admin_api, type: :request
end
```

```ruby
# {your admin request}_spec.rb

describe "..", :admin_api, type: :request do
  it "..." { foo }
end
```

**Reply by inikalaev** [reply]
Hi, I've done some tests. The above option works, but it also works without specifying :admin_api or with public_api instructions

<img width="762" alt="Снимок экрана 2024-01-12 в 14 25 37" src="https://github.com/skryukov/skooma/assets/64591049/48bfe39f-1086-4c61-ac27-24bfc0aee189">

```ruby
MyHelper = Module.new do
  def foo
    puts "foo"
  end
end

RSpec.configure do |config|
  config.include MyHelper, :admin_api, type: :request
...
end
```

I also tried options with a custom type:
```ruby
# rails_helper.rb
  Skooma.create_registry

  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema, path_prefix: "/admin/v1"], type: :admin_api

  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema, path_prefix: "/public/v1"], type: :public_api
```

```ruby
# describe "endpoint",  type: :request do
  describe "/badges", type: :admin_api do
  ...
  end
  ...
end
```
But I got the same result: admin_api endpoint not found, public_api endpoint invalid.

```ruby
  # Work
  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema], type: :request

  # Not work
  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema], type: :request
```

```ruby
  # Work
  public_schema = Rails.root.join("swagger", "public", "v1.yaml")
  config.include Skooma::RSpec[public_schema], type: :request

  # Not work
  admin_schema = Rails.root.join("swagger", "admin", "v1.yaml")
  config.include Skooma::RSpec[admin_schema], type: :request
```


```yaml
  "/admin/v1/endpoint/{id}":
    parameters:
    - name: id
      in: path
   ...

  "/public/v1/endpoint/{id}":
    parameters:
    - name: id
      in: path
   ...
```

I also tried to remove the prefixes and register them in the endpoints. In this case, the tests pass only for the scheme that was included earlier, but the second one is not found in schema.

**Reply by skryukov** [reply]
Thanks for a comprehensive response! Yes, it was `skooma` (and RSpec as well). I just merged PR with a fix: https://github.com/skryukov/skooma/pull/17 I also updated the initial answer to match the final result.

Support for multiple OpenAPI documents will be available in the next release (will publish one after fixing https://github.com/skryukov/skooma/issues/16)

Thanks again! ❤️

**Reply by inikalaev** [reply]
Hi, could you please release a version with multi-circuit support?🙏 This issue is not very critical for me because it is not a blocker, unlike multi-circuits.

**Reply by skryukov** [reply]
Sure :shipit:

**Comment by inikalaev** [comment]
Thanks!
